README.md
# Access Log Parser

## Setup

Add the library as a dependency to your `mix.exs` file:

```elixir
defp deps do
  [
    # ...
    {:accesslog_parser, "~> 0.2.0"},
    # ...
  ]
end
```

## Usage

Grab a line from your accesslog and parse it using `AccessLogParser.parse/2`:

```elixir
logline = ~s(1.2.3.4 - - [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765)
logformat = :common

AccessLogParser.parse(logline, logformat)
```

More details are available in the `AccessLogParser` module documentation.

## Benchmark

A (minimal) benchmark script is included. Please refer to the Mixfile or `mix help` output for the name.

## License

[Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
# Changelog

## v0.3.0-dev

- Enhancements
    - NCSA extended/combined log format can be parsed using either `:combined` or `:extended` format specifiers

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.9`

## v0.2.0 (2020-02-21)

- Enhancements
    - Parsing is now done using `NimbleParsec` instead of regular expressions. As a result the parser is more strict (e.g. one space between log parts instead of one-or-more) and purely numeric values like response length or status code are now returned as integers. The result map now also has atom instead of binary keys.

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.7`

## v0.1.0 (2017-04-28)

- Initial Release


test/accesslog_parser_test.exs
defmodule AccessLogParserTest do
  use ExUnit.Case, async: true

  describe "with a valid log entry" do
    # credo:disable-for-lines:6 Credo.Check.Readability.MaxLineLength
    @log_common ~s(1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765)
    @log_common_complete ~s(www.example.com 1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765 "-" "Mozilla/5.0 \(X11; Linux x86_64\) AppleWebKit/535.11 \(KHTML, like Gecko\) Chrome/17.0.963.56 Safari/535.11")
    @log_common_vhost ~s(www.example.com 1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765)
    @log_extended ~s(1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765 "-" "Mozilla/5.0 \(X11; Linux x86_64\) AppleWebKit/535.11 \(KHTML, like Gecko\) Chrome/17.0.963.56 Safari/535.11")

    @result_user_agent "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.56 Safari/535.11"

    test ":common" do
      assert AccessLogParser.parse(@log_common, :common) ==
               %{
                 date: "22/Apr/2017:15:17:39",
                 ip: "1.2.3.4",
                 length: 765,
                 method: "GET",
                 path: "/",
                 protocol: "HTTP/1.0",
                 status: 200,
                 timezone: "+0200",
                 userid: "someuser"
               }
    end

    test ":common_complete" do
      assert AccessLogParser.parse(@log_common_complete, :common_complete) ==
               %{
                 date: "22/Apr/2017:15:17:39",
                 ip: "1.2.3.4",
                 length: 765,
                 method: "GET",
                 path: "/",
                 protocol: "HTTP/1.0",
                 referrer: "-",
                 status: 200,
                 timezone: "+0200",
                 user_agent: @result_user_agent,
                 userid: "someuser",
                 vhost: "www.example.com"
               }
    end

    test ":common_vhost" do
      assert AccessLogParser.parse(@log_common_vhost, :common_vhost) ==
               %{
                 date: "22/Apr/2017:15:17:39",
                 ip: "1.2.3.4",
                 length: 765,
                 method: "GET",
                 path: "/",
                 protocol: "HTTP/1.0",
                 status: 200,
                 timezone: "+0200",
                 userid: "someuser",
                 vhost: "www.example.com"
               }
    end

    test ":combined or :extended" do
      expected = %{
        date: "22/Apr/2017:15:17:39",
        ip: "1.2.3.4",
        length: 765,
        method: "GET",
        path: "/",
        protocol: "HTTP/1.0",
        referrer: "-",
        status: 200,
        timezone: "+0200",
        user_agent: @result_user_agent,
        userid: "someuser"
      }

      assert AccessLogParser.parse(@log_extended, :combined) == expected
      assert AccessLogParser.parse(@log_extended, :extended) == expected
    end
  end

  describe "with an invalid log entry" do
    test ":common" do
      refute AccessLogParser.parse("i-am-not-valid", :common)
    end

    test ":common_complete" do
      refute AccessLogParser.parse("i-am-not-valid", :common_complete)
    end

    test ":common_vhost" do
      refute AccessLogParser.parse("i-am-not-valid", :common_vhost)
    end

    test ":extended" do
      refute AccessLogParser.parse("i-am-not-valid", :extended)
    end
  end
end


mix.exs
defmodule AccessLogParser.MixProject do
  use Mix.Project

  @url_github "https://github.com/elixytics/accesslog_parser"

  def project do
    [
      app: :accesslog_parser,
      version: "0.3.0-dev",
      elixir: "~> 1.9",
      aliases: aliases(),
      deps: deps(),
      description: "Access Log Parser",
      dialyzer: dialyzer(),
      docs: docs(),
      package: package(),
      preferred_cli_env: [
        "bench.parse": :bench,
        coveralls: :test,
        "coveralls.detail": :test
      ],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  defp aliases() do
    [
      "bench.parse": "run bench/parse.exs"
    ]
  end

  defp deps do
    [
      {:benchee, "~> 1.0", only: :bench, runtime: false},
      {:credo, "~> 1.0", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: :dev, runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:excoveralls, "~> 0.14.0", only: :test, runtime: false},
      {:nimble_parsec, "~> 1.0"}
    ]
  end

  defp dialyzer do
    [
      flags: [
        :error_handling,
        :race_conditions,
        :underspecs,
        :unmatched_returns
      ],
      plt_core_path: "plts",
      plt_file: {:no_warn, "plts/dialyzer.plt"}
    ]
  end

  defp docs do
    [
      main: "AccessLogParser",
      source_ref: "master",
      source_url: @url_github
    ]
  end

  defp package do
    %{
      files: ["CHANGELOG.md", "LICENSE", "mix.exs", "README.md", "lib"],
      licenses: ["Apache 2.0"],
      links: %{"GitHub" => @url_github}
    }
  end
end


lib/accesslog_parser.ex
# credo:disable-for-this-file Credo.Check.Readability.MaxLineLength
defmodule AccessLogParser do
  @moduledoc """
  Parses access log lines matching a given (known) format.

  ## Usage

  You simply call the parse method with your log line and the format it adheres to:

      iex> AccessLogParser.parse(~s(1.2.3.4 - - [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765), :common)
      %{
        date: "22/Apr/2017:15:17:39",
        ip: "1.2.3.4",
        length: 765,
        method: "GET",
        path: "/",
        protocol: "HTTP/1.0",
        status: 200,
        timezone: "+0200",
        userid: "-"
      }

  ## Formats

  ### Common Log Format (CLF)

  - `:common`
  - Apache configuration: `%h %l %u %t \\"%r\\" %>s %b`

  ### CLF with Virtual Host and Client

  - `:common_complete`
  - Apache configuration: `%v %h %l %u %t \\"%r\\" %>s %b \\"%{Referer}i\\" \\"%{User-agent}i\\"`

  ### CLF with Virtual Host

  - `:common_vhost`
  - Apache configuration: `%v %h %l %u %t \\"%r\\" %>s %b`

  ### NCSA extended/combined log format

  - `:combined` or `:extended`
  - Apache configuration: `%h %l %u %t \\"%r\\" %>s %b \\"%{Referer}i\\" \\"%{User-agent}i\\"`
  """

  alias AccessLogParser.Parsec

  @type log_format :: :combined | :common | :common_complete | :common_vhost | :extended

  @doc """
  Converts a given line into a map matching the defined format.
  """
  @spec parse(String.t(), log_format) :: map
  def parse(line, :combined), do: parse(line, :extended)

  def parse(line, :common) do
    case Parsec.common(line) do
      {:ok, [ip, userid, date, timezone, method, path, protocol, status, length], _, _, _, _} ->
        %{
          date: date,
          ip: ip,
          length: length,
          method: method,
          path: path,
          protocol: protocol,
          status: status,
          timezone: timezone,
          userid: userid
        }

      _ ->
        nil
    end
  end

  def parse(line, :common_complete) do
    case Parsec.common_complete(line) do
      {:ok,
       [
         vhost,
         ip,
         userid,
         date,
         timezone,
         method,
         path,
         protocol,
         status,
         length,
         referrer,
         user_agent
       ], _, _, _, _} ->
        %{
          date: date,
          ip: ip,
          length: length,
          method: method,
          path: path,
          protocol: protocol,
          referrer: referrer,
          status: status,
          timezone: timezone,
          user_agent: user_agent,
          userid: userid,
          vhost: vhost
        }

      _ ->
        nil
    end
  end

  def parse(line, :common_vhost) do
    case Parsec.common_vhost(line) do
      {:ok, [vhost, ip, userid, date, timezone, method, path, protocol, status, length], _, _, _,
       _} ->
        %{
          date: date,
          ip: ip,
          length: length,
          method: method,
          path: path,
          protocol: protocol,
          status: status,
          timezone: timezone,
          userid: userid,
          vhost: vhost
        }

      _ ->
        nil
    end
  end

  def parse(line, :extended) do
    case Parsec.extended(line) do
      {:ok,
       [
         ip,
         userid,
         date,
         timezone,
         method,
         path,
         protocol,
         status,
         length,
         referrer,
         user_agent
       ], _, _, _, _} ->
        %{
          date: date,
          ip: ip,
          length: length,
          method: method,
          path: path,
          protocol: protocol,
          referrer: referrer,
          status: status,
          timezone: timezone,
          user_agent: user_agent,
          userid: userid
        }

      _ ->
        nil
    end
  end
end


lib/accesslog_parser/parsec.ex
defmodule AccessLogParser.Parsec do
  @moduledoc false

  import NimbleParsec

  dash_or_string = choice([string("-"), ascii_string([not: 32], min: 1)])

  date = ascii_string([?a..?z, ?A..?Z, ?0..?9, ?:, ?/], 20)
  ip = ascii_string([?a..?f, ?A..?F, ?0..?9, ?:, ?.], min: 7)
  length = integer(min: 1)

  method =
    choice([
      string("CONNECT"),
      string("DELETE"),
      string("GET"),
      string("HEAD"),
      string("OPTIONS"),
      string("PATCH"),
      string("POST"),
      string("PUT"),
      string("TRACE")
    ])

  path = ascii_string([not: 32], min: 1)

  protocol =
    choice([
      string("HTTP/0.9"),
      string("HTTP/1.0"),
      string("HTTP/1.1"),
      string("HTTP/2"),
      string("HTTP/3")
    ])

  referrer = ascii_string([not: ?"], min: 1)
  status = integer(3)
  timezone = ascii_string([?0..?9, ?+, ?-], 5)
  user_agent = ascii_string([not: ?"], min: 1)
  userid = dash_or_string
  vhost = ascii_string([not: 32], min: 1)

  common =
    ip
    |> ignore(string(" "))
    |> ignore(dash_or_string)
    |> ignore(string(" "))
    |> concat(userid)
    |> ignore(string(" ["))
    |> concat(date)
    |> ignore(string(" "))
    |> concat(timezone)
    |> ignore(string(~s(] ")))
    |> concat(method)
    |> ignore(string(" "))
    |> concat(path)
    |> ignore(string(" "))
    |> concat(protocol)
    |> ignore(string(~s(" )))
    |> concat(status)
    |> ignore(string(" "))
    |> concat(length)

  common_complete =
    vhost
    |> ignore(string(" "))
    |> concat(common)
    |> ignore(string(~s( ")))
    |> concat(referrer)
    |> ignore(string(~s(" ")))
    |> concat(user_agent)

  common_vhost =
    vhost
    |> ignore(string(" "))
    |> concat(common)

  extended =
    common
    |> ignore(string(~s( ")))
    |> concat(referrer)
    |> ignore(string(~s(" ")))
    |> concat(user_agent)

  defparsec :common, common, inline: true
  defparsec :common_complete, common_complete, inline: true
  defparsec :common_vhost, common_vhost, inline: true
  defparsec :extended, extended, inline: true
end


bench/parse.exs
Benchee.run(
  %{
    ":common" => fn ->
      AccessLogParser.parse(
        ~s(1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765),
        :common
      )
    end,
    ":common_complete" => fn ->
      AccessLogParser.parse(
        ~s(www.example.com 1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765 "-" "Mozilla/5.0 \(X11; Linux x86_64\) AppleWebKit/535.11 \(KHTML, like Gecko\) Chrome/17.0.963.56 Safari/535.11"),
        :common_complete
      )
    end,
    ":common_vhost" => fn ->
      AccessLogParser.parse(
        ~s(www.example.com 1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765),
        :common_vhost
      )
    end,
    ":extended" => fn ->
      AccessLogParser.parse(
        ~s(1.2.3.4 - someuser [22/Apr/2017:15:17:39 +0200] "GET / HTTP/1.0" 200 765 "-" "Mozilla/5.0 \(X11; Linux x86_64\) AppleWebKit/535.11 \(KHTML, like Gecko\) Chrome/17.0.963.56 Safari/535.11"),
        :extended
      )
    end
  },
  formatters: [{Benchee.Formatters.Console, comparison: false}],
  warmup: 2,
  time: 10
)
